### DISCUSSION TITLE
Fail a copier renovate update when merge conflicts are present

### DISCUSSION THREAD

**Discussion by PJ-Schulz** [renovatebot/renovate#31592] [Suggest an Idea]
### Tell us more.

A merge conflict can occur during a `copier update` if a manual change has been made between two versions. I have reproduced this behavior in an example repository. This is the PR Renovate creates: https://github.com/PJ-Schulz/my-renovate-test/pull/3

In this snipped you can see the example merge conflict:

```
      containers:
      - name: my-app
<<<<<<< before updating
        image: nginx:alpine
=======
        image: nginx:1.27.1
>>>>>>> after updating
        resources:
          limits:
``` 

Currently Renovate only writes a comment in the Pull Request. In my opinion, it would be better if there was an artifact update error and the update was interpreted as failed. This way a person can manually resolve the conflict.

**Comment by rarkins** [comment]
Please rewrite this above to remove all mention of merging or automerging because that is not the root problem. Please also remove automerge from the reproduction repo and make the config _minimal_. You have 30+ lines of config of which I think 30 are probably unnecessary.

The reproduction should be a repo which has a PR open demonstrating the type of conflict.

The problem to solve here is that Renovate should mark such conflicts as artifact update errors and not "absorb" them.

**Reply by PJ-Schulz** [reply]
done

**Comment by lkubb** [comment]
I wrote the code like this intentionally because\* it allows to quickly resolve conflicts (eg via the web UI) instead of having to do everything manually. When automerging, one must have a CI check for merge conflict markers though. In general, I'm not sure if Copier updates should be automerged at all, they are much more brittle than regular dependency updates.

Failing the update instead of just warning about conflicts means I have to perform the following additional steps:
* checkout the repository locally
* create a branch
* run copier update manually
* create a PR after fixing the conflicts
* possibly request another reviewer if branch protection is active and I can't/don't want to override it

\* in addition to glossing over erroneous merge conflicts, which could still be handled with your proposal

**Reply by rarkins** [reply]
Yes, it's an important consideration that failing checks will mean that non-admins in a repo can't merge, even if they fix the reason for the failing check

**Reply by reitzig** [reply]
The inline format copier uses is problematic in other ways, too, because some Git clients (e.g. IDEA) _detect_ it as indicating a conflict, but can't then _handle_ them because no Git revs are given. Meh.

Using conflict files through `--conflict rej` (cf [docs](https://copier.readthedocs.io/en/stable/updating/#updating-a-project)) may be the "better" overall solution:

- 👍️ No "fake" Git conflict syntax in files, doesn't confuse tools.
- 👍️ Clear indicator for whether there were conflicts (be it for Renovate or the user's CI job to pick up).
- 👎️ Less intuitive to resolve for the user.
- 👎️ We wouldn't usually want to _commit_ `.rej` files

Not sure whether changes in copier (which doesn't really seem to have been designed automated updates in mind) or Renovate would be the right way to go. That said, being able to switch between the values for `--conflict` through Renovate config may be a feasible addition.

**Reply by PJ-Schulz** [reply]
> I wrote the code like this intentionally because* it allows to quickly resolve conflicts (eg via the web UI) instead of having to do everything manually.

In general I agree with this, resoliving the conflicts in the web UI is more convenient, than cloning the repo, checkout the renovate branch, resolve conflicts and push.

> In general, I'm not sure if Copier updates should be automerged at all, they are much more brittle than regular dependency updates.

This is our use case:

We use Copier to manage Kubernetes manifests on ~1000 single-node clusters. Each cluster has its own repository, where we run Renovate to update our container image versions along with other Kubernetes manifests, such as services, configmaps, ingress, ... .

Our Copier templates consists of multiple Kubernetes manifests. Our commitment is that within a major version, there are no breaking changes in the template, allowing for automatic updates. Renovate makes it easy to disable automerge for major versions to prevent unexpected breaking changes.

However, manual changes are sometimes necessary for specific clusters. These modifications are done after Copier generates the manifests, which can lead to merge conflicts during future updates. If such conflicts are merged into the main branch, our deployment pipeline gets broken.

**Comment by reitzig** [comment]
In principle, I second this. Having something like this in my (generated, hence the templating) build is not very resilient:

```bash
if    find {{ repository_start_path_prefix }}* | xargs grep -E '<{7} before updating' \
   || find {{ repository_start_path_prefix }}* | xargs grep -E '>{7} after updating'
then
    echo -e "🔥 It looks like there are unresolved conflicts left over from 'copier update'!"
    exit 1
fi
```

False positives are just waiting to happen. For instance, _documenting_ this behaviour in the (also generated) README _without_ triggering needed hacks, and limiting the check to just the templated directory took extra effort as well.
